src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs

```diff
@@ -215,63 +239,70 @@ void HandlerUpdate(object sender, EventArgs e)
+                session = _vsTestConsole.RunTestsWithCustomTestHostAsync(tests.GetGuids().Select(t => _context.VsTests[t].Case),
+                    runSettings,
                     options, eventHandler, strykerVsTestHostLauncher);
             }
 
-            // Wait for test completed report
-            if (!eventHandler.WaitEnd(timeOut+VsTestWrapperTimeOutInMs) || !session.Wait(VsTestWrapperTimeOutInMs))
+            // do we have a timeout provided ?
+            if (timeOut.HasValue)
             {
-
-                // VsTestWrapper aborts the current test sessions on timeout, except on critical error, so we have an internal timeout (+ grace period)
-                // to detect and properly handle those events. 
-                if (!strykerVsTestHostLauncher.IsProcessCreated)
+                if (!eventHandler.WaitEnd(timeOut + VsTestWrapperTimeOutInMs) ||
+                    !session.Wait(VsTestWrapperTimeOutInMs))
                 {
-                    // VsTestWrapper did not launch a test session for some reason
-                    _logger.LogError($"{RunnerId}: VsTest did not start properly.");
+                    // VsTestWrapper aborts the current test sessions on timeout, except on critical error, so we have an internal timeout (+ grace period)
+                    // to detect and properly handle those events. 
+                    if (!strykerVsTestHostLauncher.IsProcessCreated)
+                    {
+                        // VsTestWrapper did not launch a test session for some reason
+                        _logger.LogError($"{RunnerId}: VsTest did not start properly.");
+                    }
+                    else
+                    {
+                        // VsTestHost appears stuck and can't be aborted
+                        _logger.LogError(
+                            $"{RunnerId}: VsTest did not report the end of test session in due time, it may have hang.");
+                        _vsTestConsole.AbortTestRun();
+                    }
+
+                    vsTestFailed = true;
                 }
-                else
+                else if (strykerVsTestHostLauncher.ErrorCode > 0)
                 {
-                    // VsTestHost appears stuck and can't be aborted
-                    _logger.LogError($"{RunnerId}: VsTest did not report the end of test session in due time, it may have hang.");
-                    _vsTestConsole.AbortTestRun();
+                    _logger.LogError($"{RunnerId}: Test session ended with code {strykerVsTestHostLauncher.ErrorCode}");
                 }
-                _vsTestFailed = true;
-
             }
-            else if (strykerVsTestHostLauncher.ErrorCode > 0)
+            else
             {
-                _logger.LogError($"{RunnerId}: Test session ended with code {strykerVsTestHostLauncher.ErrorCode}");
+                // hard wait
```

Comment:
Are you sure this is a good idea? By far I don't have much understanding in the VSTest stuff, that's just paranoia from dev experience - having random problems because of hard waiting.

Maybe some very gracious timeout could be safer? Or, well, at least comment justifying the decision :)

Reply:
hummm
I think I will comment. But I would gladly hear what a gracious timeout could be?
Note that nothing have actually changed here, I am merely restoring the historical behavior, that has been lost because I introduced a grace delay, hence transforming a null value to a 2 seconds timeout.

Reply:
but it made me think of how the design should be improved

Reply:
Well that probably depends on possible code flows that can lead here. I mean, I just wanted to share the fear that some other edge case project will fall here and things will hang forever. Not that I know what exactly.

What's the expected time to wait here, if you have any idea? For example, if usually it'd take a few seconds to wait, we could add a timeout of 5 mins and then throw something like "Hey, something weird is going on with the project X, report this on GitHub".

Reply:
yep, I reverted to the original implementation, which was delivering the expected results. I introduced changes as an attempt to work around the 'no test' problem, discovering later than nothing could be done. But I forgot to restore the original code.